homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -0,0 +1,172 @@
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_TRAINS): [{
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_TO): cv.string,
+        vol.Required(CONF_FROM): cv.string,
+        vol.Optional(CONF_TIME): cv.time,
+        vol.Optional(CONF_WEEKDAY, default=WEEKDAYS):
+            vol.All(cv.ensure_list, [vol.In(WEEKDAYS)])}]
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(
+        hass, config, async_add_devices, discovery_info=None):
+    """Set up the departure sensor."""
+    from pytrafikverket import TrafikverketTrain
+    httpsession = async_get_clientsession(hass)
+    train_api = TrafikverketTrain(httpsession, config.get(CONF_API_KEY))
+    sensors = []
+    for train in config.get(CONF_TRAINS):
+        try:
+            from_station = yield from train_api.async_get_train_station(
+                train.get(CONF_FROM))
+            to_station = yield from train_api.async_get_train_station(
+                train.get(CONF_TO))
+        except ValueError as station_error:
+            _LOGGER.error("Problem when trying station %s to %s. Error: %s ",
+                          train.get(CONF_FROM), train.get(CONF_TO),
+                          station_error)
+            return
+
+        sensor = TrainSensor(train_api,
+                             train.get(CONF_NAME),
+                             from_station,
+                             to_station,
+                             train.get(CONF_WEEKDAY),
+                             train.get(CONF_TIME))
+        sensors.append(sensor)
+
+    async_add_devices(sensors, update_before_add=True)
+
+
+def next_weekday(fromdate, weekday):
+    """Return the date of the next time a specific weekday happen."""
+    days_ahead = weekday - fromdate.weekday()
+    if days_ahead <= 0:
+        days_ahead += 7
+    return fromdate + timedelta(days_ahead)
+
+
+def next_departuredate(departure):
+    """Calculate the next departuredate from an array input of short days."""
+    today_date = date.today()
+    today_weekday = date.weekday(today_date)
+    if WEEKDAYS[today_weekday] in departure:
+        return today_date
+    for day in departure:
+        next_departure = WEEKDAYS.index(day)
+        if next_departure > today_weekday:
+            return next_weekday(today_date, next_departure)
+    return next_weekday(today_date, WEEKDAYS.index(departure[0]))
+
+
+class TrainSensor(Entity):
+    """Contains data about a train depature."""
+
+    def __init__(self, train_api, name,
+                 from_station, to_station, weekday, time):
+        """Initialize the sensor."""
+        self._train_api = train_api
+        self._name = name
+        self._from_station = from_station
+        self._to_station = to_station
+        self._weekday = weekday
+        self._time = time
+        self._state = None
+        self._departure_state = None
+        self._delay_in_minutes = None
+
+    @asyncio.coroutine
```

Comment:
We've migrated to `async`/`await` syntax. Please drop this decorator and prefix async methods with `async` like `async def async_setup_platform`. Replace any `yield from` with `await`.